### ISSUE TITLE
SetPaketCommand takes half a second

### ISSUE THREAD

**Issue by AlexSugak** [fsprojects/Paket#3836]
### Description

An empty dotnet core 3.1 web app project takes ~1 sec to build 2nd time (no changes). Half of this 1 sec is taken by `SetPaketCommand`, even when `--no-restore` flag is set. Should this target be executed at all in this case?

### Repro steps

1. With dotnet core SDk 3.1, `cd test-proj && dotnet new web`.
2. Configure paket as a local dotnet tool, add dependencies and references files.
3. paket install, restore etc.
4. dotnet build .
5. dotnet build . --no-restore /clp:PerformanceSummary

The output has this:

```
Target Performance Summary:
        ...
        0 ms  Compile                                    1 calls
        0 ms  AfterResolveReferences                     1 calls
        0 ms  CoreBuild                                  1 calls
        0 ms  BeforeResGen                               1 calls
        ...
        0 ms  Build                                      1 calls
        ...
        3 ms  CoreCompile                                1 calls
        ...
       19 ms  PaketRestore                               1 calls
       36 ms  ResolveProjectReferences                   1 calls
       37 ms  GenerateBuildDependencyFile                1 calls
       38 ms  ProcessFrameworkReferences                 1 calls
       52 ms  FindReferenceAssembliesForReferences       1 calls
       53 ms  GenerateBuildRuntimeConfigurationFiles     1 calls
       55 ms  ResolveAssemblyReferences                  1 calls
       67 ms  ResolvePackageAssets                       1 calls
      461 ms  SetPaketCommand                            1 calls
...
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.18
```

### Expected behavior

`SetPaketCommand` should not take half the build time, preferably it should not be executed at all if `--no-restore` flag is used

### Actual behavior

`SetPaketCommand` takes ~0.5 sec, which is ~50% of build time.

### Known workarounds

N/A

**Comment by forki** [comment]
Every idea of how to improve it would be appreciated

Alexandr Sugak <notifications@github.com> schrieb am Sa., 25. Apr. 2020,
22:05:

**Comment by AlexSugak** [comment]
I ended up adding `RunPaket` variable (true by default) to `Paket.Restore.targets` and modifying  `Condition` of corresponding targets to skip running it if RunPaket is false. 

So now I can do `dotnet build . --no-restore --no-dependencies /p:RunPaket=false` and have build time < 1 sec.

Will Paket override the targets file at some point though?

My use case is that I want to do `fswatch` on .cs files and run build+test every time something changes. To make it work, the feedback time (time between a file change and a test result) must be small. This is why I am trying to squish as much time out of the build+test pipeline as possible. I don't need to do any package restores here.

**Comment by forki** [comment]
Yes Paket is adjusting the targets file with every update of itself. So
please send a PR to the one here in the embedded folder. That would ensure
that we don't break you

Alexandr Sugak <notifications@github.com> schrieb am So., 26. Apr. 2020,
13:53:

**Comment by inosik** [comment]
With your repro steps `SetPaketCommand` only takes 1-3 ms on my machine. Can you run the build with the `-bl` parameter and check if this block is executed or not:

https://github.com/fsprojects/Paket/blob/375923784ec2799f2573f0786cb7a2905d5f7067/src/Paket.Core/embedded/Paket.Restore.targets#L76-L78

You can look at the binary log with the [MSBuild Binary and Structured Log Viewer.](https://msbuildlog.com/)
